src/coreclr/tools/Common/TypeSystem/Interop/IL/Marshaller.Aot.cs

```diff
@@ -1125,4 +1127,67 @@ protected override void EmitCleanupManaged(ILCodeStream codeStream)
             codeStream.Emit(ILOpcode.call, emitter.NewToken(helper));
         }
     }
+
+    class CustomTypeMarshaller : Marshaller
+    {
+        protected override void TransformManagedToNative(ILCodeStream codeStream)
+        {
+            var marshallerType = MarshalAsDescriptor.CustomMarshallerType;
+            EmitMarshallerCreation(codeStream, marshallerType);
+            var getInstanceMethod = marshallerType.GetKnownMethod(
+                "MarshalManagedToNative",
+                new MethodSignature(MethodSignatureFlags.None, 0, Context.GetWellKnownType(WellKnownType.IntPtr), new[] { Context.GetWellKnownType(WellKnownType.Object) }));
+
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            LoadManagedValue(codeStream);
+            codeStream.Emit(ILOpcode.callvirt, emitter.NewToken(getInstanceMethod));
+            StoreNativeValue(codeStream);
+        }
+
+        protected override void TransformNativeToManaged(ILCodeStream codeStream)
+        {
+            var marshallerType = MarshalAsDescriptor.CustomMarshallerType;
+            EmitMarshallerCreation(codeStream, marshallerType);
+            var getInstanceMethod = marshallerType.GetKnownMethod(
+                "MarshalNativeToManaged",
+                new MethodSignature(MethodSignatureFlags.None, 0, Context.GetWellKnownType(WellKnownType.Object), new[] { Context.GetWellKnownType(WellKnownType.IntPtr) }));
+
+            ILEmitter emitter = _ilCodeStreams.Emitter;
+            LoadManagedValue(codeStream);
+            codeStream.Emit(ILOpcode.callvirt, emitter.NewToken(getInstanceMethod));
+            StoreNativeValue(codeStream);
+        }
+
+        protected override void EmitCleanupManaged(ILCodeStream codeStream)
+        {
+            throw new NotSupportedException();
+        }
+
+        private void EmitMarshallerCreation(ILCodeStream codeStream, TypeDesc marshallerType)
+        {
+            TypeDesc customMarshallerType = null;
+            foreach (var interfaceType in marshallerType.RuntimeInterfaces)
+            {
+                if (interfaceType.GetFullName() == "System.Runtime.InteropServices.ICustomMarshaler")
```

Comment:
This should check that the type is castable to ICustomMarshaler from the system library instead of comparing by name.

(It is what CoreCLR does.)

Reply:
Also, the methods that actually do the marshaling should be called on the ICustomMarshaler  interface; not by name.